tutorials/appengine-pusher/index.md

```diff
@@ -0,0 +1,450 @@
+To learn more about authentication in Pusher, see [Pusher's documentation](https://pusher.com/docs/authenticating_users).
+
+[embedmd]:# (java/src/main/java/com/example/appengine/pusher/AuthorizeServlet.java /public class/ $)
+```java
+public class AuthorizeServlet extends HttpServlet {
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+    // Instantiate a pusher connection
+    Pusher pusher = PusherService.getDefaultInstance();
+    // Get current logged in user credentials
+    User user = UserServiceFactory.getUserService().getCurrentUser();
+
+    // redirect to homepage if user is not authorized
+    if (user == null) {
+      response.sendRedirect("/");
+      return;
+    }
+    String currentUserId = user.getUserId();
+    String displayName = user.getNickname().replaceFirst("@.*", "");
+
+    String query = CharStreams.toString(request.getReader());
+    // socket_id, channel_name parameters are automatically set in the POST body of the request
+    // eg.socket_id=1232.12&channel_name=presence-my-channel
+    Map<String, String> data = splitQuery(query);
+    String socketId = data.get("socket_id");
+    String channelId = data.get("channel_name");
+
+    // Presence channels (presence-*) require user identification for authentication
+    Map<String, String> userInfo = new HashMap<>();
+    userInfo.put("displayName", displayName);
+
+    // Inject custom authentication code for your application here to allow/deny current request
+
+    String auth =
+        pusher.authenticate(socketId, channelId, new PresenceUser(currentUserId, userInfo));
+    // if successful, returns authorization in the format
+    //    {
+    //      "auth":"49e26cb8e9dde3dfc009:a8cf1d3deefbb1bdc6a9d1547640d49d94b4b512320e2597c257a740edd1788f",
+    //      "channel_data":"{\"user_id\":\"23423435252\",\"user_info\":{\"displayName\":\"John Doe\"}}"
+    //    }
+
+    response.getWriter().append(auth);
+  }
+
+  private static Map<String, String> splitQuery(String query) throws UnsupportedEncodingException {
+    Map<String, String> query_pairs = new HashMap<>();
+    String[] pairs = query.split("&");
+    for (String pair : pairs) {
+      int idx = pair.indexOf("=");
+      query_pairs.put(
+          URLDecoder.decode(pair.substring(0, idx), "UTF-8"),
+          URLDecoder.decode(pair.substring(idx + 1), "UTF-8"));
+    }
+    return query_pairs;
+  }
+}
+```
+
+### Send messages over a channel
+
+Authenticated clients can submit messages to the channel over HTTP using a server-side endpoint
+as shown in the following example.
+
+The sender can be [excluded](https://pusher.com/docs/server_api_guide/server_excluding_recipients)
+from receiving the broadcast message by passing in its own socket ID when triggering an event.
+Individual messages are limited to 10KB in size.
+
+For more on publishing messages to multiple channels or batching multiple messages, refer to
```

Comment:
"For more on publishing messages to multiple channels or batching multiple messages, refer to [this page](https://pusher.com/docs/server_api_guide/interact_rest_api#publishing-events)."

Change to:

"For more on publishing messages to multiple channels or batching multiple messages, refer to Pusher's [Publishing events](https://pusher.com/docs/server_api_guide/interact_rest_api#publishing-events) documentation."